src/engine/strat_engine/pool.rs

```diff
@@ -420,51 +390,43 @@ impl Pool for StratPool {
         Ok(bdev_paths)
     }
 
-    /// Precondition: All filesystems belonging to this pool must be
-    /// unmounted.
     fn destroy(self) -> EngineResult<()> {
-        teardown_pool!(self);
+        let dm = try!(DM::new());
+
+        try!(self.thin_pool.teardown(&dm));
+
         try!(self.block_devs.destroy_all());
         Ok(())
     }
 
     fn destroy_filesystems<'a, 'b>(&'a mut self,
                                    fs_uuids: &[&'b FilesystemUuid])
                                    -> EngineResult<Vec<&'b FilesystemUuid>> {
-        for fsid in fs_uuids {
-            try!(self.mdv.rm_fs(fsid));
+        let dm = try!(DM::new());
+
+        let mut removed = Vec::new();
+        for uuid in fs_uuids {
+            try!(self.thin_pool.destroy_filesystem(&dm, uuid));
+            removed.push(*uuid);
         }
-        destroy_filesystems!{self; fs_uuids}
+
+        Ok(removed)
     }
 
     fn rename_filesystem(&mut self,
                          uuid: &FilesystemUuid,
                          new_name: &str)
                          -> EngineResult<RenameAction> {
-        let old_name = rename_filesystem_pre!(self; uuid; new_name);
-
-        let mut filesystem =
-            self.filesystems
-                .remove_by_uuid(uuid)
-                .expect("Must succeed since self.filesystems.get_by_uuid() returned a value");
-
-        filesystem.rename(new_name);
-        if let Err(err) = self.mdv.save_fs(&filesystem) {
-            filesystem.rename(&old_name);
-            self.filesystems.insert(filesystem);
-            Err(err)
-        } else {
-            self.filesystems.insert(filesystem);
-            Ok(RenameAction::Renamed)
-        }
+        self.thin_pool.rename_filesystem(uuid, new_name)
     }
 
     fn rename(&mut self, name: &str) {
         self.name = name.to_owned();
     }
 
     fn get_filesystem(&mut self, uuid: &FilesystemUuid) -> Option<&mut Filesystem> {
-        get_filesystem!(self; uuid)
+        self.get_strat_filesystem(uuid)
```

Comment:
Right. Must as well get rid of macro entirely, in that case. Since SimPool and StratPool have diverged it now serves no purpose.